era5_monthly_averages_single_level_1979_2024/process_downloads.jl

```diff
@@ -0,0 +1,319 @@
+        "expver",
+        input_ds["expver"][:],
+        ("time",);
+        attrib = input_ds["expver"].attrib,
+    )
+    defVar(
+        output_ds,
+        "msuwlwrf",
+        Float32.(
+            reverse(
+                input_ds["msdwlwrf"][:, :, :] .- input_ds["msnlwrf"][:, :, :],
+                dims = 2,
+            )
+        ),
+        ("longitude", "latitude", "time"),
+        attrib = (
+            "units" => "W m**-2",
+            "long_name" => "Mean surface upward long-wave radiation flux",
+        ),
+        deflatelevel = 9,
+    )
+    defVar(
+        output_ds,
+        "msuwswrf",
+        Float32.(
+            reverse(
+                input_ds["msdwswrf"][:, :, :] .- input_ds["msnswrf"][:, :, :],
+                dims = 2,
+            )
+        ),
+        ("longitude", "latitude", "time"),
+        attrib = (
+            "units" => "W m**-2",
+            "long_name" => "Mean surface upward short-wave radiation flux",
+        ),
+        deflatelevel = 9,
+    )
+
+    defVar(
+        output_ds,
+        "latitude",
+        reverse(input_ds["latitude"][:]),
+        ("latitude",);
+        attrib = delete!(copy(input_ds["latitude"].attrib), "stored_direction"),
+    )
+
+    defVar(
+        output_ds,
+        "longitude",
+        input_ds["longitude"][:],
+        ("longitude",);
+        attrib = input_ds["longitude"].attrib,
+    )
+
+    new_times = map(input_ds["date"][:]) do t
+        d = DateTime(t)
+        d + (Day(15) - Day(d))
+    end
+
+    # check that there are no duplicates and that it is sorted
+    @assert issorted(new_times)
+    for i = 2:length(new_times)
+        @assert new_times[i] != new_times[i-1]
+    end
+    new_times_attribs = ["standard_name" => "time", "long_name" => "Time"]
+
+    defVar(output_ds, "time", new_times, ("time",); attrib = new_times_attribs)
+    close(output_ds)
+end
+
+convert_to_f32(x) = Float32(x)
+convert_to_f32(x::Missing) = x
+
+"""
+    process_hourly_data(input_ds, output_path)
+
+Processes the `input_ds` dataset to add the `msuwlwrf` and `msuwswrf` variables, which are the upward
+long-wave and short-wave radiation fluxes, respectively.
+"""
+function process_hourly_data(input_ds, output_path)
+    if isfile(output_path)
+        rm(output_path)
+        @info "Removed existing file $output_path"
+    end
+
+    output_ds = NCDataset(output_path, "c")
+    defDim(output_ds, "longitude", input_ds.dim["longitude"])
+    defDim(output_ds, "latitude", input_ds.dim["latitude"])
+    ignore_mod_31 = [1, 3, 5, 7, 9, 11, 13]
+    time_indx = filter(i -> !(i % 31 in ignore_mod_31), 1:length(input_ds["time"][:]))
+    defDim(output_ds, "time", length(time_indx))
+    missing_indices = filter(i -> (i % 31 in ignore_mod_31), 1:length(input_ds["time"][:]));
+    for index in missing_indices
+        if !all(input_ds["tcw"][:,:,index] .=== missing)
+            @error "The index pattern of the invalid data is not as expected"
+        end
+    end
```

Comment:
What does this block of code do?

Reply:
Downloading Monthly averages by hour of day results in some weirdness with the time dimension, where there are duplicate entries. Some variables are defined on one of the duplicates and some on the other. This removed the duplicate points which only hold missing data. I added a comment describing this.